invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,193 @@
+package invoices
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/clock"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/queue"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+// invoiceExpiry holds and invoice's payment hash and its expiry. This
+// is used to order invoices by their expiry for cancellation.
+type invoiceExpiry struct {
+	PaymentHash lntypes.Hash
+	Expiry      time.Time
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
+func (e invoiceExpiry) Less(other queue.PriorityQueueItem) bool {
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
+// invoices.
+// Upon start InvoiceExpiryWatcher will subscribe to all invoice events and
+// will add all active (not yet settled and not yet cancelled) invoices to its
+// watcing queue.
+// When a new invoice is added to the InvoiceRegistry, InvoiceExpiryWatcher
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
```

Comment:
Does this struct need to be publicly exported? Similarly, looking above in `htlcswitch`, it seems instead `now` (w/e that is) can be passed in, with the registry itself taking care of initialization of this struct.

Reply:
As InvoiceExpiryWatcher is a dependency of InvoiceRegistry, it needs to be public. Private initialization of dependencies afaik is not recommended as it's not leading to testable code. I think this is a pretty good article on the subject: https://testing.googleblog.com/2008/08/by-miko-hevery-so-you-decided-to.html